zetta_utils/augmentations/misalign.py

```diff
@@ -0,0 +1,162 @@
+from zetta_utils.layer.volumetric.index import VolumetricIndex
+
+T = TypeVar("T", torch.Tensor, dict[str, Any])
+
+
+@builder.register("SlipMisalign")
+@typechecked
+@attrs.mutable
+class MisalignProcessor(JointIndexDataProcessor[T, VolumetricIndex]):
+    """
+    Minimum and maximum displacement is specified in unit.
+    The selected displacement will be rounded to always be divisible
+    by `disp_in_unit_must_be_divisible_by`, which should correspond to
+    the lowest resolution of the layers being cropped.
+
+    Data handling is implemented through recording the fraction
+    of the misalignment relative to the total bbox size.
+    """
+
+    prob: float
+    disp_min_in_unit: float
+    disp_max_in_unit: float
+    disp_in_unit_must_be_divisible_by: float | None = None
+    mode: Literal["slip", "step"] = "slip"
+    keys_to_apply: list[str] | None = None
+
+    prepared_disp_fraction: tuple[float, float] | None = attrs.field(init=False, default=None)
+
+    def process_index(
+        self, idx: VolumetricIndex, mode: Literal["read", "write"]
+    ) -> VolumetricIndex:
+        if mode != "read":
+            raise NotImplementedError()
+        disp_x_in_unit_magn = random.uniform(self.disp_min_in_unit, self.disp_max_in_unit)
+        disp_y_in_unit_magn = random.uniform(self.disp_min_in_unit, self.disp_max_in_unit)
+        if self.disp_in_unit_must_be_divisible_by is not None:
+            disp_x_in_unit_magn = math.floor(
+                disp_x_in_unit_magn / self.disp_in_unit_must_be_divisible_by
+            )
+            disp_y_in_unit_magn = math.floor(
+                disp_y_in_unit_magn / self.disp_in_unit_must_be_divisible_by
+            )
+        disp_x_in_unit = disp_x_in_unit_magn * random.choice([1, -1])
+        disp_y_in_unit = disp_y_in_unit_magn * random.choice([1, -1])
+
+        disp_x_in_idx_res = disp_x_in_unit / idx.resolution[0]
+        disp_y_in_idx_res = disp_y_in_unit / idx.resolution[1]
+
+        # self.prepared_disp = (disp_x, disp_y)
+
+        start_offset = [0, 0, 0]
+        end_offset = [0, 0, 0]
+        if disp_x_in_idx_res > 0:
+            start_offset[0] = -disp_x_in_idx_res
+        else:
+            end_offset[0] = -disp_x_in_idx_res
+        if disp_y_in_idx_res > 0:
+            start_offset[1] = -disp_y_in_idx_res
+        else:
+            end_offset[1] = -disp_y_in_idx_res
+
+        idx = idx.translated_start(Vec3D[int](*start_offset)).translated_end(
+            Vec3D[int](*end_offset)
+        )
+        self.prepared_disp_fraction = (
+            disp_x_in_idx_res / idx.shape[0],
+            disp_y_in_idx_res / idx.shape[0],
+        )
+        return idx
+
+    def process_data(self, data: T, mode: Literal["read", "write"]) -> T:
+        if mode != "read":
+            raise NotImplementedError()
+
+        if isinstance(data, torch.Tensor):
+            z_size = data.shape[-1]
+        else:
+            if not self.keys_to_apply:
```

Comment:
Misalignment is a "global" augmentation, meaning it should be applied to all input/output volumes in a typical use case. Requiring users to specify the list of keys can be error-prone. Therefore, I think it's better to default to applying it to all keys.

Reply:
Step misalignment is global, but slip should only be seen through so it doesn't affect all layers. Making default behaviors differ for slip and step is misleading. This behavior was chosen because explicit is generally preferable to implicit.

Reply:
The "see through" mode for the slip misalignment is not the default; it's just an option. 

I would still argue that requiring users to specify the list of keys can be highly inconvenient and error-prone. The key point is as follows: it is highly preferable to make data augmentation modular and independent of other components of the experiment. For instance, it is very common in practice to add or remove target layers (labels) from experiment to experiment, especially when it comes to multi-task learning, e.g., embeddings, affinities, mitochondria, semantic segmentation, myelin, etc. Requiring users to specify the list of keys creates an interdependency between the data augmentation and other components, so whenever you experiment with different multi-task learning setups, you have to manually modify the list of keys every time. This is the situation that I would most like to avoid in practice.

I think the cleanest solution here would be to make `keys_to_apply` optional.

Reply:
All of the target layers can be defined in a list in a CUE file as a variable, and used in both places so that the user doesn't have to duplicate everything. Making `keys_to_apply` optional will make for a scenario where a slip missalignment will be applied to all layers. They user may not realize it, and the network will be trained with a bad setting, which could be an extremely difficult error to notice after the fact. I think explicit-over-implicit is cleaner here -- the user should specify what layer they want to shift explicitly, as it is important.

Reply:
The alternative is to make behaviors differ for slip vs step. I think it is acceptable. As long as the explicit keys are required for slip